### ISSUE TITLE
New behaviour of droplevels in Rdevel

### ISSUE THREAD

**Issue by gavinsimpson** [vegandevs/vegan#193]
R Core have altered the behaviour of `droplevels()`. For some initial discussions see #192. The basic issue is that now `droplevels` is defined as `factor(x, ..., exclude = NULL)` and this coupled with changes to `factor()` when `exclude = NULL` mean behaviour may have changed. To maintain behaviour of R v3.3 we should use `droplevels(x, exclude = NA)`. This issue exists to track changes to the range of functions in *vegan* that use `droplevels`.

Current list of functions known to use `droplevels`

- [x] `centroids.cca`
- [x] `envfit.default`
- [x] `factorfit`
- [x] `betadisper` (Fixed in ae67eba, see #192)
- [x] `hiersimu.default`
- [x] `adipart.default`
- [x] `multipart.default`

I'm tracking this on the **issue-#193** [branch](https://github.com/vegandevs/vegan/tree/issue-%23193).

**Comment by jarioksa** [comment]
It seems that `droplevels` has not been stabilized yet, but [discussion continues](https://stat.ethz.ch/pipermail/r-devel/2016-August/073012.html). Adapting to **R** change looks like trying to reach a moving target.

**Comment by gavinsimpson** [comment]
Right; I was in two minds whether to continue this when I saw the R-devel thread. However, the proposed changes don't break anything nor rely on whatever behaviour RCore ultimately settles upon, but they do make explicit the behaviour we want/expect/anticipate and so should be robust to RCore tweaking.

I would suggest we don't push out a 2.4-1 just for this though if we don't continue to generate errors when R-devel version settles down. We can. Hold off till a future 2.5-0 release if that is preferable?

**Comment by jarioksa** [comment]
It seems that `droplevels()` behaves now similarly as earlier, and `vegan_2.4-0` passes `R CMD check` without problems. The change was made in this commit:
```
r71129 | maechler | 2016-08-22 16:40:07 +0300 (Mon, 22 Aug 2016) | 1 line
Changed paths:
   M /trunk/src/library/base/R/factor.R
   M /trunk/src/library/base/man/droplevels.Rd
   M /trunk/tests/reg-tests-1c.R

droplevels() w/ smarter `exclude` default, now compatible to `[, drop=TRUE`
```
From that point of view, we need not push a new minor release. However, there are other bug fixes that may call for a minor release:

*  `goodness.cca` failed with rank=1 models and had lost `choices` argument. This also radiated to fixes in some other functions with similar problem, and the work still continues here. Typical case is that we cast a vector to a diagonal matrix with `diagmat <- diag(vec)` which fails if `vec` has only one element (`diag` is one developer biting function that can take a scalar, vector or a matrix as its first argument -- and works differently in all these cases). These bugs were introduced in 2.4-0 release to `goodness.cca`, but in some other cases they have been around for ages. Other affected functions are `stressplot.rda`, `stressplot.cca`, `fitted.dbrda`, `inertcomp`, and the search continues. It also turned out that `goodness.cca` needed fixes in several other parts. 
* `as.mcmc.oecosimu` failed when there is only one chain (that is, usually). Function is rarely used, and the bug was introduced in 2.4-0 release when `as.mcmc` gained the ability handle multiple chains.
* `nestredbetajac` and `nestedbetasor` did not cope with quantitative input data. 
* although **R** again works with `betadisper` and other `droplevels()` utilizing functions, the fixes made improve robustness and should be published.
* Current `R-devel` warns on GNU extensions in Fortran files and gives `NOTE` in **CRAN** test results. I have now fixed all these warnings in `cepin.f`, `decorana.f` and `monoMDS.f`. None of these fixes changes the behaviour of these functions: they just remove the warnings.
* There are some new features that could be safely ported: `monoMDS` gained `points` and `text` after a query in StackOverflow, `diversity` checks its input and works with `NA` , `ordixyplot` can draw polygons, `tabasco` is more configurable. In addition, there is a non-merged pull request #184 that Borcard requested for the next edition of their book.

**Comment by gavinsimpson** [comment]
Merged into `master` after merging `typo`. Closing.
